gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java

```diff
@@ -619,6 +591,81 @@ public static void runWorkUnits(String jobId, String containerId, List<WorkUnit>
     return tasks;
   }
 
+  /**
+   * Cleanup the left-over staging data possibly from the previous run of the job that may have failed
+   * and not cleaned up its staging data.
+   *
+   * Property {@link ConfigurationKeys#CLEANUP_STAGING_DATA_PER_TASK} controls whether to cleanup
+   * staging data per task, or to cleanup entire job's staging data at once.
+   *
+   * Staging data will not be cleaned if the job has unfinished {@link CommitSequence}s.
+   */
+  private void cleanLeftoverStagingData(List<WorkUnit> workUnits, JobState jobState) throws JobException {
+
+    try {
+      if (this.jobContext.getSemanticsType() == DeliverySemantics.Type.EXACTLY_ONCE
+          && this.jobContext.getCommitSequenceStore().get().exists(jobState.getJobName())) {
+        LOG.error("Job " + jobState.getJobName() + " has unfinished commit sequences. Will not clean up staging data.");
+        return;
+      }
+    } catch (IOException e) {
+      throw new JobException("Failed to clean up staging data", e);
```

Comment:
The log message should be something like `Failed to check unfinished commit sequences`.